src/c2ffi/c2ffi.lisp

```diff
@@ -0,0 +1,163 @@
+
+(defun run-program* (program args &key (output (if *trace-c2ffi* *standard-output* nil))
+                                    (error-output (if *trace-c2ffi* *error-output* nil))
+                                    ignore-error-status)
+  (when *trace-c2ffi*
+    (format *debug-io* "~&; Invoking: ~A~{ ~A~}~%" program args))
+  (zerop (nth-value 2 (uiop:run-program (list* program args) :output output :error-output error-output :ignore-error-status ignore-error-status))))
+
+(defun generate-spec-with-c2ffi (input-header-file output-spec-path &key arch sys-include-paths ignore-error-status)
+  "Run c2ffi on `INPUT-HEADER-FILE`, outputting to `OUTPUT-FILE` and
+`MACRO-OUTPUT-FILE`, optionally specifying a target triple `ARCH`."
+  (uiop:with-temporary-file (:pathname tmp-macro-file
+                             :keep *trace-c2ffi*)
+    nil
+    :close-stream
+    (let* ((arch (when arch (list "-A" arch)))
+           (sys-include-paths (loop
+                                :for dir :in sys-include-paths
+                                :append (list "-i" dir))))
+      ;; Invoke c2ffi to emit macros into TMP-MACRO-FILE
+      (when (run-program* *c2ffi-executable* (list* (namestring input-header-file)
+                                                    "-D" "null"
+                                                    "-M" (namestring tmp-macro-file)
+                                                    (append arch sys-include-paths))
+                          :output *standard-output*
+                          :ignore-error-status ignore-error-status)
+        ;; Write a tmp header file that #include's the input file and the macros file.
+        (uiop:with-temporary-file (:stream tmp-include-file-stream
+                                   :pathname tmp-include-file
+                                   :keep *trace-c2ffi*)
+          (format tmp-include-file-stream "#include \"~A\"~%" input-header-file)
+          (format tmp-include-file-stream "#include \"~A\"~%" tmp-macro-file)
+          :close-stream
+          ;; Invoke c2ffi again to generate the final output.
+          (run-program* *c2ffi-executable* (list* (namestring tmp-include-file) "-o" (namestring output-spec-path)
+                                                  (append arch sys-include-paths))
+                        :output *standard-output*
+                        :ignore-error-status ignore-error-status))))))
+
+(defun spec-path (base-name &key version (arch (local-arch)))
+  (check-type base-name pathname)
+  (make-pathname :defaults base-name
+                 :name (strcat (pathname-name base-name)
+                               (if version
+                                   (strcat "-" version)
+                                   "")
+                               "."
+                               arch)
+                 :type "spec"))
+
+(defun find-local-spec (base-name &key (otherwise nil otherwise?))
```

Comment:
it's used right here a few lines below.

i can just convert it to an ```&key error?``` or ```errorp``` or ```error-p```...

what's your preference?

FTR, this otherwise construct is crippled anyways... a full implementation is e.g. in this file: http://hub.darcs.net/hu.dwim/hu.dwim.util/browse/source/miscellaneous.lisp#249